### ISSUE TITLE
semantic version inconsistency

### ISSUE THREAD

**Issue by martey** [celery/celery#5956]
The [documentation](https://docs.celeryproject.org/en/4.4.0/contributing.html#versions) claims that Celery follows semantic versioning:

> Version numbers consists of a major version, minor version and a release number. Since version 2.1.0 we use the versioning semantics described by SemVer: http://semver.org. 

However, Celery 4.4 introduced breaking changes that are incompatible with Celery 4.3 (http://docs.celeryproject.org/en/latest/whatsnew-4.4.html#upgrading-from-celery-4-3, #5890).

I think a project governance decision needs to be made as to whether Celery plans to continue semantic versioning, or whether minor versions are allowed to have breaking changes.

**Comment by auvipy** [comment]
4.4.0 is a major version

**Comment by martey** [comment]
With semantic versioning, 4.0.0 is a major version. 4.4.0 is a minor version. From the specification at https://semver.org/ (which is explicitly linked to in the Celery documentation):

> A normal version number MUST take the form X.Y.Z where X, Y, and Z are non-negative integers, and MUST NOT contain leading zeroes. X is the major version, Y is the minor version, and Z is the patch version.

> Major version X (X.y.z | X > 0) MUST be incremented if any backwards incompatible changes are introduced to the public API.

You can decide that minor versions of Celery can include breaking changes, but it's confusing and not user friendly to claim to use semantic versioning while ignoring its most important rule (that breaking changes only happen with major versions).

**Comment by martey** [comment]
Following up on this...

I created this issue because I think it's problematic and not user-friendly for Celery to claim to use semantic versioning (creating the expectation that breaking changes will only appear in major versions like 4.0.0, as opposed to "minor" versions like 4.4.0) while not following the standard. I think there are two possible solutions to this problem.

One would be updating the documentation and the wiki to stop referencing semver and make it clear that minor versions (e.g. 4.x releases) can include breaking changes.

The other would be to actually follow semver and limit breaking changes to major versions (e.g. Celery 5.x). I think this is the better solution, because it would help people maintaining applications that use Celery and its constituent libraries know when breaking changes occur.

Other than the [initial adoption of semver 10 years ago](https://github.com/celery/celery/issues/171), the only mention of Celery's versioning structure by a core contributor outside of this issue that I have been able to find is https://github.com/celery/kombu/issues/1016. [There @thedrow wrote](https://github.com/celery/kombu/issues/1016#issuecomment-469319729):

> We don't follow SemVer to the letter since we want Celery 5 to contain certain features and signify a new path for the project.

I don't think Celery's current users care if the "new path" version of Celery is Celery 5, 6, or 7. As someone maintaining multiple Celery projects, knowing whether a new release might include breaking changes is more important to me than its explicit version number. In the past, I have checked Celery's documentation, noticed that semantic versioning was stated to be used, and then been bitten by a breaking change.

As I stated in my original issue description, this is a project governance issue. The core team / the release managers needs to decide whether they want to actually use semver or whether minor versions can have breaking changes. I can and am willing to create a pull request to update the documentation (so it stops claiming that celery uses semver), but I don't create Celery releases and don't know if this will make life more difficult for you.

**Comment by dejlek** [comment]
"I don't think Celery's current users..." - You can think whatever you like... I belong to that "current Celery's users" group and as much as I am aware of the issues, I am also aware of how difficult it is to manage project like Celery. Once you maintain a project that has few active developers and tens of thousands of users you tell me how easy is to satisfy everyone and follow principles to the letter! It is easy to complain. How about you roll up your sleeves and participate in the project, instead of bubbling here about stuff? Show some respect and gratitude! 

You do not like how the project is governed? Go find a better alternative!

**Comment by auvipy** [comment]
> "I don't think Celery's current users..." - You can think whatever you like... I belong to that "current Celery's users" group and as much as I am aware of the issues, I am also aware of how difficult it is to manage project like Celery. Once you maintain a project that has few active developers and tens of thousands of users you tell me how easy is to satisfy everyone and follow principles to the letter! It is easy to complain. How about you roll up your sleeves and participate in the project, instead of bubbling here about stuff? Show some respect and gratitude!
> 
> You do not like how the project is governed? Go find a better alternative!

I thank you wholeheartedly :)

**Comment by auvipy** [comment]
I would appreciate a PR to remove the word semver from docs. thanks.

**Comment by diefans** [comment]
> Go find a better alternative!

@dejlek A customer reports a bug and you send him to your competitors?
Fantastic marketing!

**Comment by auvipy** [comment]
> > Go find a better alternative!
> 
> @dejlek A customer reports a bug and you send him to your competitors?
> Fantastic marketing!

who is the customer here? it's a BSD licensed project.

**Comment by diefans** [comment]
> who is the customer here?

@auvipy the one that used your product and expected certain things on behalf of your communication and failed. And because he is invested in your product and its success, he pays back with contributions like reporting a bug. So you not only loose a silent customer but also somebody who is actively contributing, I guess, you especially underestimate the role of contributions like this one, since you say it is "...bubbling here about stuff? Show some respect and gratitude!"

Respect and gratitude? Blaming someone for helping you? So humble....?

**Comment by auvipy** [comment]
to be very frank and humble, I won't consider any non-paying user as a customer. sorry about that. and for the report, I clarified my statement as open to update the documentation. But I didn't like the tone so I remain silent. I didn't mean to offend you. It was just a question, not pinching or disrespecting anyone. As a non-native English speaker, my English is not perfect. I will be working on that besides improving the celery as well. While reporting issues and participating in discussions are really useful, there are more important priorities for both the core team and the community, so I would humbly request all to check the celery/ceps open PR. and consider contributing financially https://opencollective.com/celery for my personal time in celery.

**Comment by martey** [comment]
I think the comment by @dejlek is both inaccurate (I maintain an open source library with few developers and thousands of users, I have contributed code to Celery in the past, I regularly contribute code to multiple open source projects) and contains multiple unwarranted personal attacks on my character. I would even argue it violates the Code of Conduct: https://docs.celeryproject.org/en/latest/contributing.html#be-respectful

@auvipy, I found it extremely disappointing that a project maintainer would give such a comment their "wholehearted" thanks. It didn't help when you apologized to @diefans while simultaneously mentioning that you "didn't like [my] tone". Both you and @dejlek seem to be assuming malice or rudeness on my part that I don't think is evident in the words in my initial report or its followup comments. I wanted to have a discussion about improving how Celery does versions. While my personal preference would be celery adopting semver, all three of my comments mention that the status quo would be fine if the Celery documentation does not explicitly reference semver.

Under such circumstances (I'm used to personal attacks in open source, but personal attacks with the assent of the maintainer are beyond the pale), I have to retract my offer to sending a pull request to update the documentation. I still think this is a bug, so I am not closing this issue.
